scope.go

```diff
@@ -0,0 +1,92 @@
+package rcmgr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+type ResourceScope struct {
+	sync.Mutex
+
+	limit   Limit
+	memory  int64
+	buffers map[interface{}][]byte
+}
+
+var _ network.ResourceScope = (*ResourceScope)(nil)
+
+func (rc *ResourceScope) ReserveMemory(size int) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if rc.memory+int64(size) > rc.limit.GetMemoryLimit() {
+		return fmt.Errorf("cannot reserve memory: %w", ErrResourceLimitExceeded)
+	}
+
+	rc.memory += int64(size)
+	return nil
+}
+
+func (rc *ResourceScope) ReleaseMemory(size int) {
+	rc.Lock()
+	defer rc.Unlock()
+
+	rc.memory -= int64(size)
```

Comment:
Should we have a panic here if `rc.memory < 0`?

Reply:
probably a good idea, it would be a bug up the stack.

Reply:
added check with panic.